File: pegasusX/apps/warehouse-app-ios/WarehouseApp/Views/DemandForecast/ForecastSkuTable.swift
import SwiftUI

struct ForecastSkuTable: View {
    let products: [DemandForecastProduct]
    let generatedAt: String?
    let forecastDays: Int

    var body: some View {
        VStack(spacing: LabTheme.spacingLG) {
            ForEach(Array(products.enumerated()), id: \.element.id) { index, product in
                ForecastProductRow(product: product)
                    .staggeredAppear(index: index + 3)
            }

            if let generatedAt = generatedAt, !generatedAt.isEmpty {
                Text(L10n.format("mobile_warehouse.ui.generated_formattedgeneratedat_forecastdays_day_window", "\(formattedGeneratedAt(generatedAt))", "\(forecastDays)"))
                    .font(.caption2)
                    .foregroundStyle(.tertiary)
            }
        }
    }

    private func formattedGeneratedAt(_ raw: String) -> String {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        if let date = formatter.date(from: raw) ?? ISO8601DateFormatter().date(from: raw) {
            return date.formatted(date: .abbreviated, time: .shortened)
        }
        return raw
    }
}

private struct ForecastProductRow: View {
    let product: DemandForecastProduct

    var body: some View {
        VStack(alignment: .leading, spacing: LabTheme.spacingSM) {
            HStack(alignment: .firstTextBaseline) {
                Text(product.productName.isEmpty ? String(product.productId.prefix(8)) : product.productName)
                    .font(.headline)
                Spacer()
                Text(product.priority)
                    .font(.caption.bold())
                    .foregroundStyle(priorityColor(product.priority))
            }

            HStack {
                metricColumn(title: "Stock", value: "\(product.currentStock)")
                metricColumn(title: "Rec.", value: "\(product.recommendedQty)")
                metricColumn(
                    title: "Stockout",
                    value: String(format: "%.1fd", product.daysUntilStockout),
                    valueColor: stockoutColor(product.daysUntilStockout)
                )
            }

            HStack(spacing: LabTheme.spacingMD) {
                sourceChip(label: "In", value: "\(product.sources.incomingOrders)")
                sourceChip(label: "AI", value: "\(product.sources.aiPrediction)")
                sourceChip(label: "Pre", value: "\(product.sources.preOrders)")
                sourceChip(label: "Burn", value: String(format: "%.1f", product.sources.burnRate))
            }

            if let confidence = parseForecastConfidence(product.demandBreakdown) {
                ForecastConfidenceView(confidence: confidence, compact: true)
            }
        }
        .labCard()
    }

    private func metricColumn(title: String, value: String, valueColor: Color = LabTheme.label) -> some View {
        VStack(alignment: .leading, spacing: 2) {
            Text(title)
                .font(.caption2)
                .foregroundStyle(.secondary)
            Text(value)
                .font(.subheadline.monospacedDigit())
                .foregroundStyle(valueColor)
        }
        .frame(maxWidth: .infinity, alignment: .leading)
    }

    private func sourceChip(label: String, value: String) -> some View {
        VStack(spacing: 2) {
            Text(label)
                .font(.caption2)
                .foregroundStyle(.secondary)
            Text(value)
                .font(.caption.monospacedDigit())
        }
        .frame(maxWidth: .infinity)
    }

    private func priorityColor(_ priority: String) -> Color {
        switch priority.uppercased() {
        case "CRITICAL": return LabTheme.destructive
        case "URGENT": return LabTheme.warning
        default: return LabTheme.secondaryLabel
        }
    }

    private func stockoutColor(_ days: Double) -> Color {
        if days < 2 { return LabTheme.destructive }
        if days < 5 { return LabTheme.warning }
        return LabTheme.label
    }
}

File: pegasusX/apps/warehouse-app-android/app/src/main/java/com/pegasusx/warehouse/ui/screens/forecast/ForecastSkuTable.kt
package com.pegasusx.warehouse.ui.screens.forecast

import androidx.compose.ui.res.stringResource

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.grid.GridItemSpan
import androidx.compose.foundation.lazy.grid.LazyGridScope
import androidx.compose.foundation.lazy.grid.items
import androidx.compose.material3.ElevatedCard
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.style.TextOverflow
import com.pegasusx.warehouse.data.model.DemandForecastProduct
import com.pegasusx.warehouse.ui.components.ForecastConfidenceView
import com.pegasusx.warehouse.ui.theme.PegasusSpacing
import com.pegasusx.warehouse.util.parseForecastConfidence
import java.text.NumberFormat
import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.util.Locale
import com.pegasusx.warehouse.R

fun LazyGridScope.forecastSkuTable(
    products: List<DemandForecastProduct>,
    fmt: NumberFormat,
    generatedAt: String?,
    forecastDays: Int
) {
    items(products, key = DemandForecastProduct::productId) { product ->
        ForecastProductCard(product = product, fmt = fmt)
    }
    generatedAt?.takeIf { it.isNotBlank() }?.let { generated ->
        item(span = { GridItemSpan(maxLineSpan) }) {
            Text(
                stringResource(R.string.mobile_warehouse_ui_generated_formatgeneratedat_forecastdays_day_window, formatGeneratedAt(generated), forecastDays),
                style = MaterialTheme.typography.labelSmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
        }
    }
}

@Composable
private fun ForecastProductCard(
    product: DemandForecastProduct,
    fmt: NumberFormat,
) {
    val displayName = product.productName.ifBlank {
        product.productId.take(8)
    }
    ElevatedCard(Modifier.fillMaxWidth()) {
        Column(
            Modifier.padding(PegasusSpacing.lg),
            verticalArrangement = Arrangement.spacedBy(PegasusSpacing.sm),
        ) {
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Text(
                    displayName,
                    style = MaterialTheme.typography.titleSmall,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                    modifier = Modifier.weight(1f),
                )
                Text(
                    product.priority,
                    style = MaterialTheme.typography.labelMedium,
                    color = priorityColor(product.priority),
                )
            }
            Row(modifier = Modifier.fillMaxWidth()) {
                ForecastMetricColumn(
                    title = stringResource(R.string.portal_nav_stock),
                    value = fmt.format(product.currentStock),
                    modifier = Modifier.weight(1f),
                )
                ForecastMetricColumn(
                    title = stringResource(R.string.mobile_warehouse_ui_rec),
                    value = fmt.format(product.recommendedQty),
                    modifier = Modifier.weight(1f),
                )
                ForecastMetricColumn(
                    title = stringResource(R.string.warehouse_portal_forecast_forecast_sku_table_text_stockout),
                    value = String.format(Locale.US, "%.1fd", product.daysUntilStockout),
                    valueColor = stockoutColor(product.daysUntilStockout),
                    modifier = Modifier.weight(1f),
                )
            }
            Row(modifier = Modifier.fillMaxWidth()) {
                SourceChip("In", fmt.format(product.sources.incomingOrders), Modifier.weight(1f))
                SourceChip("AI", fmt.format(product.sources.aiPrediction), Modifier.weight(1f))
                SourceChip("Pre", fmt.format(product.sources.preOrders), Modifier.weight(1f))
                SourceChip(
                    "Burn",
                    String.format(Locale.US, "%.1f", product.sources.burnRate),
                    Modifier.weight(1f),
                )
            }
            parseForecastConfidence(product.demandBreakdown)?.let { confidence ->
                ForecastConfidenceView(confidence = confidence, compact = true)
            }
        }
    }
}

@Composable
private fun ForecastMetricColumn(
    title: String,
    value: String,
    modifier: Modifier = Modifier,
    valueColor: Color = MaterialTheme.colorScheme.onSurface,
) {
    Column(modifier = modifier) {
        Text(title, style = MaterialTheme.typography.labelSmall, color = MaterialTheme.colorScheme.onSurfaceVariant)
        Text(value, style = MaterialTheme.typography.bodyMedium, fontFamily = FontFamily.Monospace, color = valueColor)
    }
}

@Composable
private fun SourceChip(label: String, value: String, modifier: Modifier = Modifier) {
    Column(modifier = modifier, horizontalAlignment = Alignment.CenterHorizontally) {
        Text(label, style = MaterialTheme.typography.labelSmall, color = MaterialTheme.colorScheme.onSurfaceVariant)
        Text(value, style = MaterialTheme.typography.labelMedium, fontFamily = FontFamily.Monospace)
    }
}

@Composable
private fun priorityColor(priority: String): Color = when (priority.uppercase(Locale.US)) {
    "CRITICAL" -> MaterialTheme.colorScheme.error
    "URGENT" -> MaterialTheme.colorScheme.tertiary
    else -> MaterialTheme.colorScheme.onSurfaceVariant
}

@Composable
private fun stockoutColor(days: Double): Color = when {
    days < 2 -> MaterialTheme.colorScheme.error
    days < 5 -> MaterialTheme.colorScheme.tertiary
    else -> MaterialTheme.colorScheme.onSurface
}

private fun formatGeneratedAt(raw: String): String = runCatching {
    val instant = Instant.parse(raw)
    DateTimeFormatter.ofPattern("MMM d, HH:mm")
        .withZone(ZoneId.systemDefault())
        .format(instant)
}.getOrDefault(raw)
